frontend/frontend-filters-links.php

```diff
@@ -465,8 +465,13 @@ public function check_canonical_url( $requested_url = '', $do_redirect = true )
 
 		// The language is not correctly set so let's redirect to the correct url for this object
 		if ( $do_redirect && $redirect_url && $requested_url != $redirect_url ) {
-			wp_safe_redirect( $redirect_url, 301, POLYLANG );
-			exit;
+			// Protect against chained redirects.
+			if ( $redirect_url === $this->check_canonical_url( $redirect_url, false ) ) {
+				wp_safe_redirect( $redirect_url, 301, POLYLANG );
+				exit;
+			} else {
+				return;
+			}
```

Comment:
The `return` at this position looks strange to me.

In the current situation, when `$do_redirect` is true, if we don't redirect, we always return the requested url (although the variable is named `$redirect_url`).
In your proposal, if we dont redirect, either we return nothing or we return the requested url, depending of the failing test. I would prefer only one case, for example, always return nothing:
```php
if ( $do_redirect ) {
	if ( $redirect_url && $requested_url != $redirect_url && $redirect_url === $this->check_canonical_url( $redirect_url, false ) ) {
		wp_safe_redirect( $redirect_url, 301, POLYLANG );
		exit;
	} else {
		return;
	}
}

```
and keep the return valiue only when `$do_redirect` is false. What do you think?

Reply:
Indeed closer than WordPress code. Done !